src/Pathy/Path.purs

```diff
@@ -0,0 +1,291 @@
+-- | "invalid path" that may be constructed is using the `parentOf` function,
+-- | e.g. `parentOf rootDir`, or by parsing an equivalent string such as
+-- | `/../`, but such paths may not be rendered to strings until they are first
+-- | sandboxed to some directory.
+data Path (a :: RelOrAbs) (b :: DirOrFile)
+  = Init
+  | ParentOf (Path a Dir)
+  | In (Path a Dir) (Name b)
+
+derive instance eqPath :: Eq (Path a b)
+derive instance ordPath :: Ord (Path a b)
+
+instance showPathRelDir :: (IsRelOrAbs a, IsDirOrFile b) => Show (Path a b) where
+  show p@Init = foldRelOrAbs (const "currentDir") (const "rootDir") p
+  show (ParentOf p) = "(parentOf " <> show p <> ")"
+  show (In p n) = "(" <> show p <> " </> " <> foldDirOrFile (("dir " <> _) <<< show) (("file " <> _) <<< show) n <> ")"
+
+-- | A type describing a file or directory path.
+type AnyPath a = Either (Path a Dir) (Path a File)
+
+-- | A type describing a relative file or directory path.
+type RelPath = AnyPath Rel
+
+-- | A type describing an absolute file or directory path.
+type AbsPath = AnyPath Abs
+
+-- | A type describing a directory whose location is given relative to some
+-- | other, unspecified directory (referred to as the "current directory").
+type RelDir = Path Rel Dir
+
+-- | A type describing a directory whose location is absolutely specified.
+type AbsDir = Path Abs Dir
+
+-- | A type describing a file whose location is given relative to some other,
+-- | unspecified directory (referred to as the "current directory").
+type RelFile = Path Rel File
+
+-- | A type describing a file whose location is absolutely specified.
+type AbsFile = Path Abs File
+
+-- | The root directory, which can be used to define absolutely-located resources.
+rootDir :: Path Abs Dir
+rootDir = Init
+
+-- | The "current directory", which can be used to define relatively-located
+-- | resources.
+currentDir :: Path Rel Dir
+currentDir = Init
+
+-- | Creates a path which points to a relative file of the specified name.
+file :: NonEmptyString -> Path Rel File
+file = file' <<< Name
+
+-- | Creates a path which points to a relative file of the specified name.
+file' :: Name File -> Path Rel File
+file' = In currentDir
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir :: NonEmptyString -> Path Rel Dir
+dir = dir' <<< Name
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir' :: Name Dir -> Path Rel Dir
+dir' = In currentDir
+
+-- | Creates a path that points to the parent directory of the specified path.
+parentOf :: forall a. Path a Dir -> Path a Dir
+parentOf p = ParentOf p
+
+-- | Extends a path with a file or directory under the current path.
+extendPath :: forall a b. Path a Dir -> Name b -> Path a b
+extendPath p = In p
+
+-- | Given a directory path, appends a relative path to extend the original
+-- | path.
+appendPath :: forall a b. Path a Dir -> Path Rel b -> Path a b
+appendPath = case _, _ of
+  Init, Init -> Init
```

Comment:
I didn't realise I had done! But looking at the other PR... they were only necessary as inference hints for `AppendOutcome`, there's no reason to have them otherwise, since `Init` will already infer as `Path a b` (it was required for AppendOutcome as it wouldn't be able to find an instance for `Path t0 t1` where t0/t1 are unknowns).